#9 Encoders for MQTT

sbcd90 opened the pull request.

Lazzaretti commented on cloudevents-sdk-mqtt/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+[dependencies]
+cloudevents-sdk = { version = "0.2.0", path = ".." }
+lazy_static = "1.4.0"
+paho-mqtt = { path = "../../paho.mqtt.rust" }
```
Why are you adding the dependency by path and not by cargo by default dependency?

```suggestion
paho-mqtt = {version="0.7.1"}
```

sbcd90 replied:
@Lazzaretti , the version `0.7.1` has an issue. It only allows property from head of container.

https://github.com/eclipse/paho.mqtt.rust/blob/7b3a29cdca902c3b9f0ab20909ed24bd47b6e881/src/properties.rs#L466

this is fixed in current master

https://github.com/eclipse/paho.mqtt.rust/blob/master/src/properties.rs#L900

Unfortunately, there is no release for it yet. 
According to cloudevents specification, we have to put all headers under prop code `UserProperty`.

https://github.com/cloudevents/spec/blob/master/mqtt-protocol-binding.md#314-examples

slinkydeveloper replied:
Can we get a new release of paho.mqtt soon?

sbcd90 replied:
@slinkydeveloper @Lazzaretti i currently linked it to git repo.

slinkydeveloper replied:
Still, this is not valid to perform a release of the package to creates.io

slinkydeveloper replied:
In any case, we need a release from them. @fpagliughi Are you going to perform a release of paho.mqtt any time soon?

fpagliughi replied:
Yes, apologies for the delay. Lots going on these days. I'm starting work back on Paho Rust this week. I think the strategy will be to push out what's been sitting up on GitHub for the last few monhts as v0.8, so as not to break the code people have been creating with it. That should happen before the end of the week.

Then work will start to update the underlying C lib and add some new features. That should go out by the end of the year as v0.9, and will have some minor breaking changes. So people can choose.

slinkydeveloper commented:
I marked this as 0.4, hopefully paho.mqtt new version will be released before our new version and we can get this one in. I'll review this PR asap

slinkydeveloper commented:
Seems like musl doesn't like paho.mqtt? Is there a way to fix it? I see this issue is open here: https://github.com/eclipse/paho.mqtt.rust/issues/57

One thing I wanted with the integration between sdk-rust and mqtt is a crate supporting no_std (given we're also investing a lot of time in supporting no std #94). Is there any particular reason why @sbcd90 you chose paho.mqtt? Did you considered other mqtt clients supporting no_std?

sbcd90 commented:
I did evaluate https://github.com/bytebeamio/rumqtt . I can switch to this library.
However, i went with `paho mqtt` because I was not sure of future maintenance guarantees for `rumqtt` library.

sbcd90 commented:
Please let me know if i should try switching the library.

fpagliughi commented:
A Paho release went out last week (v0.8).

Another (v0.9) is in active development and should be out by the end of the year. The initial purpose for the next release was to get HTTP/S proxy support for websocket connections, but a few other additions popped up, and now we're also going to push for better cross-compiling support and fully static linking. That includes (optional) static  OpenSSL linking and getting that musl support stable.

slinkydeveloper commented:
@sbcd90 I think we can go ahead and work on this one, but let's be explicit we're supporting paho, so maybe rename it to `cloudevents-sdk-paho-mqtt`

sbcd90 commented:
> @sbcd90 I think we can go ahead and work on this one, but let's be explicit we're supporting paho, so maybe rename it to `cloudevents-sdk-paho-mqtt`. Can you set the version to 0.8?

I have set version to `0.8`

slinkydeveloper commented on cloudevents-sdk-paho-mqtt/src/mqtt_consumer_record.rs:
```diff
@@ -0,0 +1,239 @@
+}
+
+impl ConsumerMessageDeserializer {
+    fn get_mqtt_headers(message: &Message) -> Result<HashMap<String, Vec<u8>>> {
```
Do we need the result return here? I don't see any failing stmt here

sbcd90 replied:
updated this

slinkydeveloper commented on cloudevents-sdk-paho-mqtt/src/mqtt_consumer_record.rs:
```diff
@@ -0,0 +1,239 @@
+    fn encoding(&self) -> Encoding {
+        match (
+            self.headers
+                .get("content-type")
```
Can you use the constant in headers?

sbcd90 replied:
updated this

slinkydeveloper commented on cloudevents-sdk-paho-mqtt/src/mqtt_consumer_record.rs:
```diff
@@ -0,0 +1,239 @@
+        headers::MqttVersion::V3_1_1 => {
+            StructuredDeserializer::into_event(ConsumerMessageDeserializer::new(msg)?)
+        }
+    }
```
I may miss some knowledge about mqtt here, but why do we need the headers version? Also, why in V5 only Binary Deserializer is supported (aka only binary mode is supported), more than the general MessageDeserializer?

Also, can you infer the headers version from `Message` more than providing it manually as parameter?

sbcd90 replied:
now, V5 should allow both structured & binary mode. V3 only structured mode.

slinkydeveloper commented on cloudevents-sdk-paho-mqtt/src/mqtt_consumer_record.rs:
```diff
@@ -0,0 +1,230 @@
+                .flatten()
+                .map(|s| s.starts_with(headers::CLOUDEVENTS_JSON_HEADER))
+                .unwrap_or(false),
+            self.headers
+                .get(headers::MQTT_VERSION_HEADER)
+                .map(|s| String::from_utf8(s.to_vec()).ok())
+                .flatten()
+                .map(|s| s.eq(headers::MQTT_V5_BINARY))
+                .unwrap_or(false),
```
So do we expect incoming messages to have this header?

slinkydeveloper commented on cloudevents-sdk-paho-mqtt/src/mqtt_producer_record.rs:
```diff
@@ -0,0 +1,140 @@
+
+    pub fn from_event(event: Event) -> Result<Self> {
+        match event
+            .extension(headers::MQTT_VERSION_HEADER)
```
Is this part of the spec?

slinkydeveloper reviewed (commented):
So i have some concerns about the whole v3/v5 difference, why did you defined an extension for it? do we need it?

sbcd90 commented:
> So i have some concerns about the whole v3/v5 difference, why did you defined an extension for it? do we need it?

Hi @slinkydeveloper , currently, v3 supports only structured mode. v5 supports both structured as well as binary mode according to spec.
This is because mqtt v3 has no support for headers while mqtt v5 has.
Now, in the current approach the default mode is structured mode which is supported by both v3  & v5. If binary mode needs to be used, then we need a flag to tell us if the message has to be deserialized using BinaryDeserializer.

Earlier, I was passing the flag as a parameter, now I pass it as an extension field. paho-mqtt passes version info as a param.
See https://github.com/cloudevents/sdk-rust/blob/ba09b473e329626ef635aa65c400b94356eae65e/example-projects/paho-mqtt-example/src/main.rs#L80

Kindly let me know according to you which is the best approach.

slinkydeveloper commented:
Ok so one step at the time.

### MQTT -> CloudEvent

First of all, i think you should remove the headers map you create and you should directly access to the `Property` borrow, in order to avoid allocating intermediate memory, which in this case it's not required.

For figuring out v3 vs v5 messages, I think we should simply do best effort: from the paho.mqtt rustdocs i see that you can always get the properties from a message https://docs.rs/paho-mqtt/0.8.0/paho_mqtt/message/struct.Message.html#method.properties and in case the `UserProperty` is empty, then we can assume the properties are not there so it's not mqtt v5 (hence we should just deserialize as structured).

### CloudEvent -> MQTT

In this case, we should have an additional input in the serialization that explicits the mqtt version, like an enum `MqttVersion` created by us with `MQTT_3` and `MQTT_5` variants. I thought there was such enum in paho.mqtt, but there isn't. No extensions, the event **must** not be mutated.

Also note that i don't expect the user providing `MessageRecord`, this should be done under the hood:

```
pub trait MessageBuilderExt {
    fn event(self, Event, MqttVersion) -> MessageBuilder;
}
```

And the implementation would look like:

```
impl MessageBuilderExt for MessageBuilder {
    fn event(self, Event, MqttVersion) -> MessageBuilder {
        // mutate directly MessageBuilder here using MessageRecord
    }
}
```

Check out the actix-web, because it's a better one (I see that you probably looked at the `MessageRecord` thingy from Kafka, but that's kinda bad because of the rdkafka apis themselves)

Does that makes sense? Did I understood the problems here?

sbcd90 commented:
Hi @slinkydeveloper , Thanks a lot for your detailed review. Refactored the code accordingly.
Added a sample README.

Can you please have a look now?

slinkydeveloper reviewed (commented):
To me this PR sounds good, @sbcd90 can you fix MUSL tests? Is paho-mqtt compatible with musl?

@Lazzaretti can you review this PR?

sbcd90 commented:
hi @slinkydeveloper , i fixed the remaining issues.
For musl, i get the error

```
Make sure you also have the development packages of openssl installed.
  For example, `libssl-dev` on Ubuntu or `openssl-devel` on Fedora.
```

slinkydeveloper commented:
@sbcd90 try to modify the github action here: https://github.com/cloudevents/sdk-rust/blob/master/.github/workflows/rust_tests.yml#L31 adding openssl dev package setup

slinkydeveloper commented on example-projects/paho-mqtt-example/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+serde_json = "^1.0"
+futures = "^0.3"
+tokio = { version = "^0.2", features = ["full"] }
+clap = "2.33.1"
```
Here you need to add 

```
[workspace]

```

Like https://github.com/cloudevents/sdk-rust/blob/master/example-projects/warp-example/Cargo.toml

sbcd90 commented:
hi @slinkydeveloper , the builds are fixed now.

Lazzaretti commented on cloudevents-sdk-paho-mqtt/README.md:
````diff
@@ -0,0 +1,42 @@
+- Start the MQTT V3 Consumer
+
+```
+run --package <package-name> --bin <binary-name> -- --mode consumerV3 --broker tcp://localhost:1883 --topic test
````
what is `run`? can we have some more specific examples?

Lazzaretti commented on Cargo.toml:
```diff
@@ -28,6 +28,8 @@ bitflags = "^1.2"
 
 [target."cfg(not(target_arch = \"wasm32\"))".dependencies]
 hostname = "^0.3"
+openssl-sys = "*"
+openssl = { version = "*", features = ["vendored"] }
```
may we should add some version restrictions?

slinkydeveloper replied:
Good catch! Why in the main toml we need openssl?

Lazzaretti commented on cloudevents-sdk-paho-mqtt/src/mqtt_producer_record.rs:
```diff
@@ -0,0 +1,150 @@
+        }
+    }
+
+    pub fn from_event(event: Event, version: &headers::MqttVersion) -> Result<Self> {
+        match version {
+            headers::MqttVersion::MQTT_5 => {
+                BinaryDeserializer::deserialize_binary(event, MessageRecord::new())
+            }
```
As I understood MQTTv5 should support Binary and Structured Content Mode.
https://github.com/cloudevents/spec/blob/v1.0.1/mqtt-protocol-binding.md#32-structured-content-mode

So I would add a parameter to define the content mode (Binary or Structured) and not the MQTT version directly.

slinkydeveloper replied:
More than `version` you would use the `Encoding` struct? sounds good
